File: processing/processing/dtos.py
from typing import Optional

from pydantic import BaseModel

from processing.constants import (
    NDSI_BAND_BIO,
    NDSI_BAND_ANTHRO,
    ADI_BIN_STEP,
    ADI_DB_THRESHOLD,
    ADI_IMPL,
    SPECTRO_STFT_OVERLAP_RATIO,
    SPECTRO_STFT_WINDOW_MS,
    SPECTRO_STFT_WINDOW_TYPE,
    SPECTRO_N_BANDS,
    SPECTRO_DBFS_REF,
    HT_FRAME_SIZE,
    MED_FRAME_SIZE,
    MFCC_N_MFCC,
    LEQ_PERCENTILE_VALUE,
    LEQ_DIFF_PERCENTILE_A,
    LEQ_DIFF_PERCENTILE_B,
    LEQ_SHORT_DT,
    MPS_STFT_1_WINDOW_MS,
    MPS_STFT_1_OVERLAP_RATIO,
    MPS_STFT_2_WINDOW_MS,
    MPS_STFT_2_OVERLAP_RATIO,
    MPS_N_BANDS,
    SPECTRO_SCALE,
    MPS_SCALE,
)
from processing.enums import (
    ExtractorImpl,
    ReducerImpl,
    AutoclusterImpl,
    MetricImpl,
    AdiImpl,
    FrequencyScale,
    StftWindowType,
    ComputationStrategy,
)


class FileDto(BaseModel):
    Index: str
    Path: str
    Date: str
    Site: str
    tags: dict[str, str]


class SettingsDto(BaseModel):
    storagePath: str
    audioPath: str
    expectedSampleRate: int
    timelineOrigin: str
    audioHost: str
    timezone: str
    computationStrategy: ComputationStrategy
    computationDimensions: int
    computationIterations: int
    displaySeed: int
    memoryLimit: int | float


class BandDto(BaseModel):
    index: int
    name: str
    low: int
    high: int


class IntegrationDto(BaseModel):
    index: int
    name: str
    duration: int


class ExtractorDto(BaseModel):
    index: int
    name: str
    impl: ExtractorImpl

    window: int
    hop: int

    spectro_n_bands: int = SPECTRO_N_BANDS
    spectro_scale: FrequencyScale = SPECTRO_SCALE
    spectro_stft_window_type: StftWindowType = SPECTRO_STFT_WINDOW_TYPE
    spectro_stft_window_ms: Optional[int] = SPECTRO_STFT_WINDOW_MS
    spectro_stft_overlap_ratio: float = SPECTRO_STFT_OVERLAP_RATIO
    spectro_dbfs_ref: float = SPECTRO_DBFS_REF

    mps_n_bands: int = MPS_N_BANDS
    mps_scale: FrequencyScale = MPS_SCALE
    mps_stft_1_window_ms: Optional[int] = MPS_STFT_1_WINDOW_MS
    mps_stft_1_overlap_ratio: float = MPS_STFT_1_OVERLAP_RATIO
    mps_stft_2_window_ms: Optional[int] = MPS_STFT_2_WINDOW_MS
    mps_stft_2_overlap_ratio: float = MPS_STFT_2_OVERLAP_RATIO

    mfcc_n_mfcc: int = MFCC_N_MFCC

    ndsi_band_bio: tuple[int, int] = NDSI_BAND_BIO
    ndsi_band_anthro: tuple[int, int] = NDSI_BAND_ANTHRO

    adi_bin_step: int = ADI_BIN_STEP
    adi_db_threshold: int = ADI_DB_THRESHOLD
    adi_impl: AdiImpl = ADI_IMPL

    ht_frame_size: int = HT_FRAME_SIZE
    med_frame_size: int = MED_FRAME_SIZE

    leq_percentile_dt: float = LEQ_SHORT_DT
    leq_percentile_value: int = LEQ_PERCENTILE_VALUE

    leq_diff_dt: float = LEQ_SHORT_DT
    leq_diff_percentile_a: int = LEQ_DIFF_PERCENTILE_A
    leq_diff_percentile_b: int = LEQ_DIFF_PERCENTILE_B


class ReducerDto(BaseModel):
    index: int
    impl: ReducerImpl
    dimensions: int


class RangeDto(BaseModel):
    index: int
    name: str
    start: str
    end: str


class AutoclusterDto(BaseModel):
    index: int
    impl: AutoclusterImpl
    minClusterSize: int
    minSamples: int
    alpha: float
    epsilon: float


class MetricDto(BaseModel):
    index: int
    impl: MetricImpl


class TrajectoryDto(BaseModel):
    index: int
    name: str
    start: str
    end: str
    tagName: str
    tagValue: str
    smoothingWindow: int


class ExtractionDto(BaseModel):
    index: int
    name: str
    bands: list[BandDto]
    integrations: list[IntegrationDto]
    extractors: list[ExtractorDto]
    reducers: list[ReducerDto]
    autoclusters: list[AutoclusterDto]
    metrics: list[MetricDto]
    trajectories: list[TrajectoryDto]


class JsonDto(BaseModel):
    version: str
    settings: SettingsDto
    extractions: list[ExtractionDto]
    ranges: list[RangeDto]
    files: list[FileDto]

File: shared/dtos.ts
import {INCLUDE_IN_AGGREGATION_DEFAULT} from '@shared/constants';
import {z} from 'zod';

import {
  AdiImpl,
  AutoclusterImpl,
  ComputationStrategy,
  ExtractorImpl,
  FrequencyScale,
  MetricImpl,
  ReducerImpl,
  StftWindowType,
} from './enums';

const dateString = z.string().refine((str) => !Number.isNaN(Date.parse(str)), {
  message: 'Invalid date string',
});

export const FileDto = z.object({
  Index: z.string(),
  Path: z.string(),
  Date: dateString,
  Site: z.string(),
  tags: z.record(z.string(), z.string()),
});

export type FileDto = z.infer<typeof FileDto>;

export const SettingsDto = z.object({
  storagePath: z.string(),
  audioPath: z.string(),
  expectedSampleRate: z.number().int(),
  timelineOrigin: dateString,
  audioHost: z.string(),
  timezone: z.string(),
  computationStrategy: ComputationStrategy,
  computationDimensions: z.number().int(),
  computationIterations: z.number().int(),
  displaySeed: z.number().int(),
  memoryLimit: z.number(),
});

export type SettingsDto = z.infer<typeof SettingsDto>;
export const BandDto = z.object({
  index: z.number().int(),
  name: z.string(), // unique
  low: z.number().int(),
  high: z.number().int(),
});

export type BandDto = z.infer<typeof BandDto>;

export const IntegrationDto = z.object({
  index: z.number().int(),
  name: z.string(), // unique
  duration: z.number().int(), // ms
});

export type IntegrationDto = z.infer<typeof IntegrationDto>;

export const ExtractorDto = z.object({
  index: z.number().int(),
  name: z.string(), // unique
  impl: ExtractorImpl,
  include_in_aggregation: z.boolean().default(INCLUDE_IN_AGGREGATION_DEFAULT),

  window: z.number().int(), // ms
  hop: z.number().int(), // ms

  spectro_n_bands: z.number().int().optional(),
  spectro_scale: FrequencyScale.optional(),
  spectro_stft_window_type: StftWindowType.optional(),
  spectro_stft_window_ms: z.number().int().nullable().optional(),
  spectro_stft_overlap_ratio: z.number().optional(),
  spectro_dbfs_ref: z.number().optional(),

  mps_n_bands: z.number().int().optional(),
  mps_scale: FrequencyScale.optional(),
  mps_stft_1_window_ms: z.number().int().optional(),
  mps_stft_1_overlap_ratio: z.number().optional(),
  mps_stft_2_window_ms: z.number().int().optional(),
  mps_stft_2_overlap_ratio: z.number().optional(),

  mfcc_n_mfcc: z.number().int().optional(),

  ndsi_band_bio: z.tuple([z.number().int(), z.number().int()]).optional(),
  ndsi_band_anthro: z.tuple([z.number().int(), z.number().int()]).optional(),

  adi_bin_step: z.number().int().optional(),
  adi_db_threshold: z.number().int().optional(),
  adi_impl: AdiImpl.optional(),

  ht_frame_size: z.number().int().optional(),
  med_frame_size: z.number().int().optional(),

  leq_percentile_dt: z.number().optional(),
  leq_percentile_value: z.number().int().optional(),

  leq_diff_dt: z.number().optional(),
  leq_diff_percentile_a: z.number().int().optional(),
  leq_diff_percentile_b: z.number().int().optional(),
});

export type ExtractorDto = z.infer<typeof ExtractorDto>;

export const ReducerDto = z.object({
  index: z.number().int(),
  impl: ReducerImpl,
  dimensions: z.number().int(),
});

export type ReducerDto = z.infer<typeof ReducerDto>;

export const AutoclusterDto = z.object({
  index: z.number().int(),
  impl: AutoclusterImpl,
  minClusterSize: z.number(),
  minSamples: z.number(),
  alpha: z.number(),
  epsilon: z.number(),
});

export type AutoclusterDto = z.infer<typeof AutoclusterDto>;

export const MetricDto = z.object({
  index: z.number().int(),
  impl: MetricImpl,
});

export type MetricDto = z.infer<typeof MetricDto>;

export const TrajectoryDto = z.object({
  index: z.number().int(),
  name: z.string(), // unique
  start: dateString,
  end: dateString,
  tagName: z.string(),
  tagValue: z.string(),
  smoothingWindow: z.number().int().positive(),
});

export type TrajectoryDto = z.infer<typeof TrajectoryDto>;

export const ExtractionDto = z.object({
  index: z.number().int(),
  name: z.string(),
  bands: BandDto.array(),
  integrations: IntegrationDto.array(),
  extractors: ExtractorDto.array(),
  reducers: ReducerDto.array(),
  autoclusters: AutoclusterDto.array(),
  metrics: MetricDto.array(),
  trajectories: TrajectoryDto.array(),
});

export type ExtractionDto = z.infer<typeof ExtractionDto>;

export const RangeDto = z.object({
  index: z.number().int(),
  name: z.string(), // unique
  start: dateString,
  end: dateString,
});

export type RangeDto = z.infer<typeof RangeDto>;

export const ConfigDto = z.object({
  version: z.string(),
  settings: SettingsDto,
  extractions: ExtractionDto.array(),
  ranges: RangeDto.array(),
  files: FileDto.array(),
});

export type ConfigDto = z.infer<typeof ConfigDto>;
